### PR TITLE
Handle unions in is_proper_subtype

### LINKED ISSUE #2840: assert isinstance can widen types
Here's an example:
```py
x = ""  # type: str
reveal_type(x)  # E: Revealed type is 'builtins.str'
assert isinstance(x, (int, str))
reveal_type(x)  # E: Revealed type is 'Union[builtins.int, builtins.str]'
```

We should limit `assert isinstance(...)` to strictly narrowing types instead.

### PR DESCRIPTION
**Pull request by kirbyfan64**
Fixes #2840.

If you guys don't want to change `is_proper_subtype`, I have another fix, too.

### CODE REVIEW

File: test-data/unit/check-isinstance.test [hunk 1]
```diff
@@ -534,6 +534,22 @@ if isinstance(v, (B, C)):
...
+
+a = A()  # type: A
+assert isinstance(a, (A, B))
+reveal_type(a)  # E: Revealed type is '__main__.A'
+
+b = A()  # type: Union[A, B]
+assert isinstance(b, (A, B, C))
+reveal_type(b)  # E: Revealed type is 'Union[__main__.A, __main__.B]'
+[builtins fixtures/isinstance.pyi]
+
```
Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: I would add one more test with non-overlapping types, like:
```
a = A()  # type: A
assert isinstance(a, (B, C))
reveal_type(a)
```

Comment (kirbyfan64) [inline test-data/unit/check-isinstance.test]: @ilevkivskyi What do you think the correct answer would be there? I would think that should be an error or something...

Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: I think it should be revealed as ``UninhabitedType()`` (pretty-printed as ``<uninhabited>`` by ``reveal_type()``). It could be also an error, but as I understand, we currently do not flag appearances of ``UninhabitedType`` as errors, just prohibit any operations with this type.

Comment (kirbyfan64) [inline test-data/unit/check-isinstance.test]: @ilevkivskyi Actually, I think that should be a separate issue. The fact that it currently leaves `a`'s type as `A` seems entirely unrelated to this PR and would require some more complicated changes.

Comment (kirbyfan64) [inline test-data/unit/check-isinstance.test]: @ilevkivskyi Done: https://github.com/python/mypy/pull/2841/commits/95db7ecb87d31a172000a0e8840fb074ead58e71

**Comment by kirbyfan64** [comment]
@ilevkivskyi Actually, I broke everything and evidently have no clue how to get that working...

Could this PR just be left as-is? I mean, it fixes the bug, right? ;)

**Comment by ilevkivskyi** [comment]
OK, ``UninhabitedType`` is a difficult part of mypy and maybe you right. Also this could be related to how ``Void`` vs ``NoneTyp`` vs ``UninhabitedType`` are treated.

I am not insisting here.

**Comment by JukkaL** [comment]
This seems good to me, thanks! Feel free to continue working on the non-overlapping case if you are interested, but it doesn't seem directly related.
